### PR TITLE
[DELETE with DVs] Allow reading CDF from files with Deletion Vectors - Part 2/2

### PR DESCRIPTION
**Pull request by xupefei**
## Description

This PR is part of https://github.com/delta-io/delta/issues/1701.

This is a follow-up of https://github.com/delta-io/delta/pull/1680 to add support to allow reading CDC from files that have DV associated. In this PR we modify the CDC reader to construct in-line DVs diff'ed from two existing DVs, and modify the corresponding FileIndex to use the in-line DV.

## How was this patch tested?

Adding a CDC variant to an existing suite.

### CODE REVIEW

**Comment by xupefei** [comment]
Due to https://github.com/delta-io/delta/pull/1680 not being merged, the current PR looks bigger than what it actually is. The real diff is https://github.com/delta-io/delta/pull/1704/commits/0d183a2695aba2784895bc1e3e86ba95818d9f5f.

**Comment by xupefei** [comment]
@vkorukanti Could you review?

File: core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala [hunk 1]
```diff
@@ -154,6 +157,16 @@ object CDCReader extends CDCReaderImpl
       timestamp: Timestamp,
       actions: Seq[T],
       commitInfo: Option[CommitInfo])
+
+  case class FileVersion(version: Long, timestamp: Timestamp) {
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:161]: One or two lines doc describing what this is used for. Same comment for below case class.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala [hunk 2]
```diff
@@ -608,31 +617,196 @@ trait CDCReaderImpl extends DeltaLogging {
...
-        new TahoeRemoveFileIndex(
-          spark, removeFileSpecs.toSeq, deltaLog, deltaLog.dataPath, snapshot),
-        snapshot.protocol,
-        snapshot.metadata,
-        isStreaming))
+
+    // If a path is only being removed, then scan it normally as removed rows
+    (removeFilesMap.keySet -- addFilesMap.keySet).foreach { removeKey =>
+      finalRemoveFiles
+        .getOrElseUpdate(new FileVersion(removeKey), ListBuffer())
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:675]: `FileVersion` seems like the wrong name? It looks like it is referring the the delta log version?

File: core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala [hunk 3]
```diff
@@ -608,31 +617,196 @@ trait CDCReaderImpl extends DeltaLogging {
...
+    val dfAddsAndRemoves = ListBuffer[DataFrame]()
+
+    finalAddFiles.foreach { case (fileVersion, addFiles) =>
+      val commitInfo = versionToCommitInfo.get(fileVersion.version)
+      dfAddsAndRemoves.append(
+        scanIndex(
+          spark,
+          new CdcAddFileIndex(
+            spark,
+            Seq(CDCDataSpec(
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:688]: nit: nested method for cdcDataSpec(fileVersion, addFiles, commitInfo) and use it here and below.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala [hunk 4]
```diff
@@ -751,6 +925,88 @@ trait CDCReaderImpl extends DeltaLogging {
...
+    // 2) Remove without DV, add with DV1:
+    //    Rows masked by DV1 are deleted.
+    // 3) Remove with DV1, add without DV:
+    //    Rows masked by DV1 are added. May happen when restoring a table.
+    // 4) Remove with DV1, add with DV2:
+    //   a) Rows masked by DV2 but not DV1 are deleted.
+    //   b) Rows masked by DV1 but not DV2 are re-added. May happen when restoring a table.
+    (removeDvOpt, addDvOpt) match {
+      case (None, None) =>
+        throw new Exception("Unreadable code path reached.")
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:958]: nit: add few more details to the error message: ex. same file is added and removed within the same version

File: core/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala [hunk 5]
```diff
@@ -39,7 +39,8 @@ class CdcAddFileIndex(
     filesByVersion: Seq[CDCDataSpec[AddFile]],
     deltaLog: DeltaLog,
     path: Path,
-    snapshot: Snapshot
+    snapshot: Snapshot,
+    override val rowIndexFilters: Option[Map[String, RowIndexFilterType]] = None
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala:43]: docs for the new param. Also mention the details such as the path should be a URI encoded in order for the look up to work.

Comment (vkorukanti) [reply core/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala:43]: same for remove. Also do we want to add assert to check if there is one entry for each file?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSuite.scala [hunk 6]
```diff
@@ -964,3 +964,11 @@ class DeltaCDCScalaSuite extends DeltaCDCSuiteBase {
   }
 
 }
+
+class DeltaCDCScalaWithDeletionVectorsSuite extends DeltaCDCScalaSuite
```
Comment (vkorukanti) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSuite.scala:968]: does this suite cover the streaming tests?

**Review by vkorukanti — CHANGES_REQUESTED** [review]
LGTM, few comments.

**Comment by vkorukanti** [comment]
If I understand correctly, the [DELETE with DVs](https://github.com/delta-io/delta/issues/1591) doesn't need to do anything special for CDF? If yes, please remove the CDF work item from the plan.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala (hunk 1 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:161]: Done!

File: core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala (hunk 4 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:958]: Done!

File: core/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala (hunk 5 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala:43]: Done adding docs.

Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala:43]: It's not necessary that `rowIndexFilters` contains all files in `filesByVersion`, the former is a superset of the latter. IMO asserting here won't bring too much add value.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSuite.scala (hunk 6 above)
Comment (xupefei) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSuite.scala:968]: Nope... but I just added one!

**Comment by xupefei** [comment]
> If I understand correctly, the [DELETE with DVs](https://github.com/delta-io/delta/issues/1591) doesn't need to do anything special for CDF? If yes, please remove the CDF work item from the plan.

Thanks! I addressed all your comments.
I don't have the edit permission for the [DELETE with DVs](https://github.com/delta-io/delta/issues/1591) page, could you do it instead?

**Review by vkorukanti — APPROVED** [review]

**Pull request closed without merging**